lib/instrumentation/modules/ioredis.js

```diff
@@ -1,81 +1,69 @@
 'use strict'
 
-var semver = require('semver')
+// Instrumentation of the 'ioredis' package:
+// https://github.com/luin/ioredis
+// https://github.com/luin/ioredis/blob/master/API.md
 
-var shimmer = require('../shimmer')
+const semver = require('semver')
 
-var spanSym = Symbol('elasticAPMSpan')
+const constants = require('../../constants')
+const { getDBDestination } = require('../context')
+const shimmer = require('../shimmer')
+
+const TYPE = 'cache'
+const SUBTYPE = 'redis'
+const hasIoredisSpanSym = Symbol('ElasticAPMHasIoredisSpan')
 
 module.exports = function (ioredis, agent, { version, enabled }) {
+  if (!enabled) {
+    return ioredis
+  }
   if (!semver.satisfies(version, '>=2.0.0 <5.0.0')) {
     agent.logger.debug('ioredis version %s not supported - aborting...', version)
     return ioredis
   }
 
-  agent.logger.debug('shimming ioredis.Command.prototype.initPromise')
-  shimmer.wrap(ioredis.Command && ioredis.Command.prototype, 'initPromise', wrapInitPromise)
-
-  if (!enabled) return ioredis
+  const ins = agent._instrumentation
 
   agent.logger.debug('shimming ioredis.prototype.sendCommand')
   shimmer.wrap(ioredis.prototype, 'sendCommand', wrapSendCommand)
-
   return ioredis
 
-  // wrap initPromise to allow us to get notified when the callback to a
-  // command is called. If we don't do this we will still get notified because
-  // we register a callback with command.promise.finally the
-  // wrappedSendCommand, but the finally call will not get fired until the tick
-  // after the command.callback have fired, so if the transaction is ended in
-  // the same tick as the call to command.callback, we'll lose the last span
-  // as it hasn't yet ended.
```

Comment:
REVIEW NOTE: In ioredis, `sendCommand(cmd)` is called with a `Command` object that always has a `.promise` and can optionally have a user-supplied `.callback` function. When the command is complete: (a) the callback is called, if any, and (b) the promise is resolved or rejected. Because this is a Promise that resolve/reject is in the next tick.

A change this update makes is to only instrument the promise. The code before this would instrument both, and `span.end()` on the first one. This means that a slight behaviour change with this update is that when a user-supplied callback is called, then span is not yet ended. It will be ended in the next tick with the promise. Not that the run-context fixes in this change mean that the span is intentionally *not* the `currentSpan` in the user-callback, so the slight change on end time doesn't have any impact that I can see.

The comment above about "we'll lose the last span" is wrong: the Elastic APM system can handle a trace with a span ending after the transaction ends just fine. It looks like this:

<img width="1366" alt="Screen Shot 2021-11-19 at 12 38 51 PM" src="https://user-images.githubusercontent.com/46866/142709514-8b22d312-2474-489f-b76a-4a1fb88428ad.png">

I like that the code is less complex: only needing to `command.promise.then()` to handle `span.end()`.